### PR TITLE
Add formula crystal (0.8.0)

### PR DESCRIPTION
**Pull request by waj**
Making use of the released `crystal` name at #44417 (Thanks @ctate!)

### CODE REVIEW

File: Library/Formula/crystal.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      end
+    end
+
+    bin.install ".build/crystal"
+    bin.install ".build/shards" if build.with?("shards")
+    prefix.install "src"
+  end
+
+  def test
+    system "crystal", "eval", "puts 1"
```
Comment (bfontaine) [inline Library/Formula/crystal.rb]: Please use `"#{bin}/crystal"` instead of `"crystal"`.

**Comment by DomT4** [comment]
@xu-cheng I thought we couldn't use the old names at this point as they became symlinks? Was that fixed?

File: Library/Formula/crystal.rb (hunk 1 above)
Comment (waj) [inline Library/Formula/crystal.rb]: Done

**Comment by xu-cheng** [comment]
@DomT4 if users have installed the formula using the oldname, they won't be able to install new formula(although we need to add some check in `brew install` to warn this situation) unless they uninstall old formula. For other users, it will be ok. 

Since there will be some adjustments for core logic required, let's hold off merging this PR for the moment.

**Comment by ctate** [comment]
@xu-cheng Anything I can help with? If you let me know the exact behavior, I can submit a PR.

**Comment by waj** [comment]
@xu-cheng Is there any planned timeframe for that adjustment? Should I wait? I could otherwise name this formula `crystal-lang` and rename it to `crystal` in the future when the logic change is done.
Is there any open issue I can follow to know when such change is done and I can rename the formula back?

**Comment by will** [comment]
Awesome, looking forward to this, thanks @waj

**Comment by DomT4** [comment]
`crystal-lang` seems fine to me until we decide how to handle freeing up old names properly, unless Xu has a magical fix in mind already, which he may well do :smile:.

**Comment by waj** [comment]
It's a pity that actually @ctate asked exactly about this before sending the PR (https://github.com/Homebrew/homebrew/pull/44417#issuecomment-144293862) but now neither can use the name.

I can wait if you say that the problem will be fixed in a reasonable time and we avoid another rename in the future.

**Comment by mikemcquaid** [comment]
We can't promise a timeframe as we're all volunteers working in our spare time. I do recommend just waiting, though.

**Comment by waj** [comment]
@mikemcquaid I totally understand and I don't expect any promise. Just saying that if someone is currently working in the fix then it might be better to wait. Otherwise I'll rename the formula and leave `crystal` unused.

**Comment by mikemcquaid** [comment]
No-one is working on it right now but it's a planned feature for future. Perhaps renaming this at a later stage is a good idea.

**Comment by waj** [comment]
Ok, I just renamed the formula then. Thanks!

**Comment by waj** [comment]
The build is failing but I think it's not my fault :innocent:
Anything else needed to get this through?

File: Library/Formula/crystal-lang.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        system buildpath/"bin/crystal", "build", "-o", buildpath/".build/shards", "src/shards.cr"
+      end
+    end
+
+    bin.install ".build/crystal"
+    bin.install ".build/shards" if build.with?("shards")
+    prefix.install "src"
+  end
+
+  def test
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: Please use `test do` instead of `def test`.

File: Library/Formula/crystal-lang.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
+class CrystalLang < Formula
+  homepage "http://crystal-lang.org/"
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: Please add a short description above this line using `desc "..."`.

File: Library/Formula/crystal-lang.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    sha256 "95916792766e42e3b005b144190c3f88d5cb8bcbaaddeb8fa8ced8bac1ef424d"
+  end
+
+  depends_on "libevent"
+  depends_on "libpcl"
+  depends_on "bdw-gc"
+  depends_on "llvm" => :build
+
+  option "without-release", "Do not build the compiler in release mode"
+  option "without-shards", "Do not include `shards` dependency manager"
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: Please move the `option`s before the `depends_on` block.

**Comment by waj** [comment]
@bfontaine Done

File: Library/Formula/crystal-lang.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
+class CrystalLang < Formula
+  desc "Efficient, type safe and statically compiled language with Ruby inspired syntax"
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: Please shorten this description. “`crystal-lang: <desc>` should fit under 80 chars.

```
==> audit problems
crystal-lang:
 * Description is too long. "name: desc" should be less than 80 characters.
Length is calculated as crystal-lang + desc. (currently 93)
```

Comment (waj) [inline Library/Formula/crystal-lang.rb]: Done!

**Comment by waj** [comment]
@bfontaine Anything else?

File: Library/Formula/crystal-lang.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    if build.head?
+      ENV["CRYSTAL_CONFIG_VERSION"] = `git rev-parse --short HEAD`.strip
+    else
+      ENV["CRYSTAL_CONFIG_VERSION"] = version
+    end
+
+    ENV["CRYSTAL_CONFIG_PATH"] = prefix/"src:libs"
+    ENV.append_path "PATH", "boot/bin"
+
+    if build.with?("release")
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: You can drop the parentheses: `build.with? "release"`.

File: Library/Formula/crystal-lang.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
+class CrystalLang < Formula
+  desc "Fast and statically typed, compiled language with Ruby like syntax"
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: `Ruby-like`

File: Library/Formula/crystal-lang.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    ENV["CRYSTAL_CONFIG_PATH"] = prefix/"src:libs"
+    ENV.append_path "PATH", "boot/bin"
+
+    if build.with?("release")
+      system "make", "crystal", "release=true"
+    else
+      system "make", "llvm_ext"
+      mkdir ".build"
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: Use `(buildpath/".build").mkpath`

File: Library/Formula/crystal-lang.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    if build.with?("shards")
+      resource("shards").stage do
+        system buildpath/"bin/crystal", "build", "-o", buildpath/".build/shards", "src/shards.cr"
+      end
+    end
+
+    bin.install ".build/crystal"
+    bin.install ".build/shards" if build.with?("shards")
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: You can move this in the `if build.with?("shards")` above.

File: Library/Formula/crystal-lang.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    if build.with?("release")
+      system "make", "crystal", "release=true"
+    else
+      system "make", "llvm_ext"
+      mkdir ".build"
+      system "bin/crystal", "build", "-o", ".build/crystal", "src/compiler/crystal.cr"
+    end
+
+    if build.with?("shards")
```
Comment (bfontaine) [inline Library/Formula/crystal-lang.rb]: You can drop the parentheses here too.

**Comment by bfontaine** [comment]
@waj Some style nitpicking but otherwise we’re good :smile:

**Comment by waj** [comment]
@bfontaine Done!

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

For future reference the preferred commit message format for new formulae is `crystal-lang 0.8.0 (new formula)`.

**Pull request closed without merging**